### PR TITLE
REINVT-1661 Migrate to Typescript

### PR DESCRIPTION
**Pull request by aleics-lx**
This PR migrates the current action into Typescript taking as a reference the [Github Typescript Action](https://github.com/actions/typescript-action) template.

**Enhancements**
 - Added prettier and eslint for code linting.
 - Added typescript compiler to work together with `ncc` (as stated in the [docs](https://github.com/vercel/ncc#with-typescript)).
 - Remove of custom building script `update-dist.sh` with npm script for a more "ecosystem-friendly" approach, and proper versioning of `ncc`. Also, no need that developers install `ncc` globally in their laptops.
 - Use of types in code for type-safe usage, that will allow faster development cycles and robustness.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+import * as fs from 'fs';
+import * as path from 'path';
+import { AzcopyCommand, AzureStorageAccount } from './azcopy-command';
+import { ExecOptions } from '@actions/exec/lib/interfaces';
+import { getChangedFiles } from './get-changed-files';
+import moment from 'moment';
+import simplegit from 'simple-git/promise';
+
+// eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires, import/no-commonjs
+const noopStream = require('stream-blackhole')();
```
Comment (aleics-lx) [inline src/index.ts:12]: I wasn't sure about this one, that's why I've let it as it was. But looking at the `stream-blackhole` source code, I guess we could just copy the implementation and make it "EcmaScript-friendly".

**Comment by konstantintieber** [comment]
Whoa this is awesome! This will close https://github.com/leanix/storage-deploy-action/issues/23

File: src/index.ts (hunk 1 above)
Comment (konstantintieber) [reply src/index.ts:12]: Indeed [the source](https://github.com/Floby/node-stream-blackhole/blob/master/index.js) isn't complex.

**Comment by perbernhardt** [comment]
Thanks so much @aleics-lx , I really like this! Would not put an approve on this though, as I cannot really say something on the actual implementation 😛 But LGTM!

**Comment by yae-leanix** [comment]
Agreed, awesome stuff @aleics-lx! 🚀 
I'll take a closer look after lunch :)

**Review by konstantintieber — APPROVED** [review]
The tests are green, so I'd say let's ship it!

... after doing one test deployment of pathfinder using a feature branch: https://github.com/leanix/leanix-pathfinder-web/runs/5463074203?check_suite_focus=true

I didn't compare the new `index.ts` code with the `index.js` code line by line, but I was really happy to see all the wonderful typings you added.

**Comment by aleics-lx** [comment]
> The tests are green, so I'd say let's ship it!
> 
> ~... after doing one test deployment of pathfinder using a [feature branch](https://github.com/leanix/leanix-pathfinder-web/tree/feature/REINVT-1661-test-typescript-storage-deploy-action): [Link to test deployment to my instance](https://github.com/leanix/leanix-pathfinder-web/runs/5463074203?check_suite_focus=true)~
> 
> ~[Link to commit with the workflow changes](https://github.com/leanix/leanix-pathfinder-web/commit/43da1f6d8d3e276ee133cafa9ca29b330b885ffd)~
> 
> Please disregard the first workflow test. I shouldn't have enabled version-deployment for a test deployment without providing a dedicated "test name" for the `app-name` input.
> 
> I'll trigger another one after lunch.
> 
> I didn't compare the new `index.ts` code with the `index.js` code line by line, but I was really happy to see all the wonderful typings you added.

Thanks for testing this as well from the consumer side! Forgot to do it myself 🙏  I've re-run the job, but I guess it has the same inputs as your run, so it would fail again. I'll wait for your testing then :) 

Regarding the changes in `index.ts/js`, I would advice to take a look into the commit: https://github.com/leanix/storage-deploy-action/pull/33/commits/dd4e3079cbb15d5515761d4463a8112a77ee9fc0. The global diffing is not easy since the formatting broke git's tree diffing. Same applies for all the other files :)

**Comment by konstantintieber** [comment]
> Regarding the changes in index.ts/js, I would advice to take a look into the commit: https://github.com/leanix/storage-deploy-action/commit/dd4e3079cbb15d5515761d4463a8112a77ee9fc0.

Very good tip! That makes it a lot easier :D

**Comment by konstantintieber** [comment]
Here's the new "integration testing" branch on pathfinder-web: https://github.com/leanix/leanix-pathfinder-web/tree/feature/REINVT-1661-test-new-typescript-deploy-action

[Link to test workflow run](https://github.com/leanix/leanix-pathfinder-web/runs/5463843439?check_suite_focus=true)

**Comment by aleics-lx** [comment]
@konstantintieber @yae-leanix I just realised that there was a small bug in the `build` command, which made the project to actually not be compiled properly. Now it should be fixed. Maybe the integration tests need to be triggered again, sorry for that.

**Comment by konstantintieber** [comment]
> @konstantintieber @yae-leanix I just realised that there was a small bug in the `build` command, which made the project to actually not be compiled properly. Now it should be fixed. Maybe the integration tests need to be triggered again, sorry for that.

Here's the rerun of the integration workflow run, the first one worked perfectly already: https://github.com/leanix/leanix-pathfinder-web/runs/5463843439?check_suite_focus=true

**Review by konstantintieber — APPROVED** [review]
One more time for the force push 😇

**Review by yae-leanix — APPROVED** [review]
Just awesome all around 🚀  Thanks so much for investing the work!

**Comment by konstantintieber** [comment]
First successful pathfinder-web shell app release after TypeScript migration: https://github.com/leanix/leanix-pathfinder-web/runs/5480056211?check_suite_focus=true

🥳

**Pull request merged**